pkg/clusteragent/clusterchecks/dispatcher_nodes.go

```diff
@@ -43,3 +44,53 @@ func (d *dispatcher) processNodeStatus(nodeName string, status types.NodeStatus)
 
 	return (node.lastConfigChange == status.LastChange), nil
 }
+
+// getLeastBusyNode returns the name of the node that is assigned
+// the lowest number of checks. In case of equality, one is chosen
+// randomly, based on map iterations being randomized.
+func (d *dispatcher) getLeastBusyNode() string {
+	var leastBusyNode string
+	minCheckCount := int(-1)
+
+	d.store.RLock()
+	defer d.store.RUnlock()
+
+	for name, store := range d.store.nodes {
+		if name == "" {
+			continue
+		}
+		if minCheckCount == -1 || len(store.digestToConfig) < minCheckCount {
+			leastBusyNode = name
+			minCheckCount = len(store.digestToConfig)
+		}
+	}
+	return leastBusyNode
+}
+
+// expireNodes iterates over nodes and removes the ones that have not
+// reported for more than the expiration duration. It returns configs
+// that were dispatched to these nodes, to re-dispatch them.
+func (d *dispatcher) expireNodes() []integration.Config {
```

Comment:
I think the name of this method only partially covers its scope.
It does delete the stale nodes from the store  but it also returns the orphaned configs.
Could we have something like

```
func (d *dispatcher) retrieveStaleState() ([]integration.Config, staleNodes string) {
...
~delete(d.store.nodes, name)~
staleNodes = append(staleNodes, name)
...
return orphanedConfigs, staleNodes
```
And delete the nodes there https://github.com/DataDog/datadog-agent/pull/2392/files#diff-bd226a7b7bfd2b2f5a425bae0935f3cdR92 ?

Reply:
That would indeed be more readable, but would not be atomic anymore.
Doing that, we might run into a race condition where we would schedule a config on an expired node between the retrieve and delete calls. We'd lose it, as it would not be listed in `orphanedConfigs`